homeassistant/components/elkm1/__init__.py

```diff
@@ -278,11 +273,74 @@ def _keypad_changed(keypad: Element, changeset: dict[str, Any]) -> None:
     for keypad in elk.keypads:
         keypad.add_callback(_keypad_changed)
 
+
+async def _ensure_elk_connection(elk: Elk, host: str) -> None:
+    """Ensure ElkM1 connection is established and authenticated."""
+
+    def _raise_auth_failed(message: str) -> None:
+        """Raise ConfigEntryAuthFailed with cleanup."""
+        elk.disconnect()
+        raise ConfigEntryAuthFailed(message)
+
+    def _raise_not_ready(message: str, exc: Exception) -> None:
+        """Raise ConfigEntryNotReady with cleanup."""
+        elk.disconnect()
+        raise ConfigEntryNotReady(message) from exc
```

Comment:
This creates performance-wise even an overhead, and as it's used only once, we should inline it directly